README.md
# @qiwi/common-formatters

[![CI](https://github.com/qiwi/common-formatters/actions/workflows/ci.yaml/badge.svg)](https://github.com/qiwi/common-formatters/actions/workflows/ci.yaml)
[![Maintainability](https://api.codeclimate.com/v1/badges/aa149c9058728b89a577/maintainability)](https://codeclimate.com/github/qiwi/common-formatters/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/aa149c9058728b89a577/test_coverage)](https://codeclimate.com/github/qiwi/common-formatters/test_coverage)
[![npm (scoped)](https://img.shields.io/npm/v/@qiwi/common-formatters)](https://www.npmjs.com/package/@qiwi/common-formatters)

> Common string formatters

### Motivation
There's no such thing as universal formatters collection, but ... Ha, and this one library is just a set of several controversial workarounds. 
What can be said in defense? Smaller is better.

## Usage
```javascript
    import {formatMoney, formatNumber, formatCardPan, formatPhone, formatPercent} from '@qiwi/common-formatters'
```

##### Money
```javascript
    formatMoney(12345.6789)   // '12 345,68'
    formatMoney(12300.45, {currencyCode: 'RUB', fractionDelimiter: '.'}) // '12 300.45 ₽'
    formatMoney(123.45, {currencySymbol: 'Foo'}) // '123,45 Foo'
    
    // Available opts
    {
      strict?: boolean
      digitDelimiter?: string
      fractionDelimiter?: string
      fractionLength?: number
      currencyCode?: string
      currencySymbol?: string
      currencyPosition?: 'left' | 'right'
      currencySpacer?: string
      sign?: boolean
      fractionRemoveZeros?: boolean
    }
```

##### Numbers
```javascript
        
    formatNumber(12345.6789)  //  '12 345,6789'
    formatNumber(12345.6789, {digitDelimiter: ',', fractionDelimiter: '.'}) // '12,345.6789'
    
    // Opts
    {
      digitDelimiter: string;
      fractionDelimiter: string;
      fractionLength?: number;
      strict: boolean;
      sign: boolean;
    }
```

##### Percents
```javascript
        
    formatPercent(0.123)  //  '12,30%'
    formatPercent(567, {sign: true, fractionLength: 0}) // '56 700%'
    
    // Opts
    {
      strict: boolean;
      digitDelimiter: string;
      fractionDelimiter: string;
      fractionLength: number;
      multiplier: number;
      sign: boolean;
    }
```

##### Phone numbers
* [E.164](https://en.wikipedia.org/wiki/E.164)
* [Formatting International Phone Numbers](https://support.twilio.com/hc/en-us/articles/223183008-Formatting-International-Phone-Numbers)
```javascript
    // Basic cases
    formatPhone('1234567')     // 123-45-67
    formatPhone('12345678')    // 1234-5678
    formatPhone('12345')       // 1-23-45
    formatPhone('1234567890')  // 1234567890
    
    // Format by mask
    formatPhone('79876543210', {mask: '+* *** ***-**-**'}) // +7 987 654-32-10
    
    // Format by opts
    formatPhone('223344', {countryCode: '7', areaCode: '8443', areaBrackets: true, phoneNumberDelimiter: '_'}) // +7 (8443) 22_33_44

    // Opts
    {
        blocksDelimiter: string;
        countryCode: string;
        areaCode: string;
        areaBrackets: boolean;
        areaCodeLength: number;
        countryCodePrefix: string;
        countryCodeLength: number;
        phoneNumberLength: number;
        phoneNumberDelimiter: string;
        mask: string;
    }
```

##### Card PAN
```javascript
    formatCardPan('1234567812345678', {digitDelimiter: '-'}) // '1234-5678-1234-5678'
```

### License
[MIT](./LICENSE)


CHANGELOG.md
## [1.2.3](https://github.com/qiwi/common-formatters/compare/v1.2.2...v1.2.3) (2023-08-06)

### Fixes & improvements
* perf: up deps ([243f32c](https://github.com/qiwi/common-formatters/commit/243f32cfeea72ec403ca1104fad3e8959dbbb6f8))

## [1.2.2](https://github.com/qiwi/common-formatters/compare/v1.2.1...v1.2.2) (2023-03-18)

### Fixes & improvements
* fix: sync version ([6286c13](https://github.com/qiwi/common-formatters/commit/6286c130445fd6b02e6aceb6d9f0a45734b3062c))
* refactor: apply strict maven dir layout ([cdcb9fd](https://github.com/qiwi/common-formatters/commit/cdcb9fd37e6c6a1d1ed3d5f005a146f37d194735))

## [1.0.0](https://github.com/qiwi/common-formatters/compare/undefined...v1.0.0) (2023-03-18)

### Fixes & improvements
* refactor: apply strict maven dir layout ([cdcb9fd](https://github.com/qiwi/common-formatters/commit/cdcb9fd37e6c6a1d1ed3d5f005a146f37d194735))

# [1.2.0](https://github.com/qiwi/common-formatters/compare/v1.1.0...v1.2.0) (2021-03-09)


### Bug Fixes

* **package:** up deps, fix vuls ([1dc9060](https://github.com/qiwi/common-formatters/commit/1dc906086f0d6c59553d13496af54d537268b96a))


### Features

* **money:** provide currency positioning flag ([5627b72](https://github.com/qiwi/common-formatters/commit/5627b7299931afd668b49705b867b0580484617d))

# [1.1.0](https://github.com/qiwi/common-formatters/compare/v1.0.0...v1.1.0) (2020-11-05)


### Features

* removing zeros in fractional parts ([#65](https://github.com/qiwi/common-formatters/issues/65)) ([bfdc884](https://github.com/qiwi/common-formatters/commit/bfdc8841d7cb6547ffe9d6f5d913bea96f870ea4))

# [1.0.0](https://github.com/qiwi/common-formatters/compare/v0.7.9...v1.0.0) (2020-11-05)


### Features

* non-breaking space between value and symbol ([015ece3](https://github.com/qiwi/common-formatters/commit/015ece3698eace858eab36d0df87c8a71d94f14c))


### Performance Improvements

* **package:** deps revision ([73d858f](https://github.com/qiwi/common-formatters/commit/73d858fb5f209cdab2a5baf4271ce5b413d11568))


### BREAKING CHANGES

* **package:** require tslib v2

# [0.8.0](https://github.com/qiwi/common-formatters/compare/v0.7.9...v0.8.0) (2020-11-05)


### Features

* non-breaking space between value and symbol ([015ece3](https://github.com/qiwi/common-formatters/commit/015ece3))

## [0.7.9](https://github.com/qiwi/common-formatters/compare/v0.7.8...v0.7.9) (2019-08-20)


### Performance Improvements

* **package:** deps revision ([bd6dc83](https://github.com/qiwi/common-formatters/commit/bd6dc83))

## [0.7.8](https://github.com/qiwi/common-formatters/compare/v0.7.7...v0.7.8) (2019-05-27)


### Bug Fixes

* redeclare facade deps ([81f3f79](https://github.com/qiwi/common-formatters/commit/81f3f79))

## [0.7.7](https://github.com/qiwi/common-formatters/compare/v0.7.6...v0.7.7) (2019-05-27)


### Bug Fixes

* **build:** rebuild as commonjs ([9bbc101](https://github.com/qiwi/common-formatters/commit/9bbc101))

## [0.7.6](https://github.com/qiwi/common-formatters/compare/v0.7.5...v0.7.6) (2019-05-27)


### Bug Fixes

* import money pack as dir ([c310561](https://github.com/qiwi/common-formatters/commit/c310561))

## [0.7.5](https://github.com/qiwi/common-formatters/compare/v0.7.4...v0.7.5) (2019-03-07)


### Performance Improvements

* tech release ([3ac1c49](https://github.com/qiwi/common-formatters/commit/3ac1c49))

## [0.7.4](https://github.com/qiwi/common-formatters/compare/v0.7.3...v0.7.4) (2019-03-07)


### Performance Improvements

* tech release ([caedf50](https://github.com/qiwi/common-formatters/commit/caedf50))

## [0.7.3](https://github.com/qiwi/common-formatters/compare/v0.7.2...v0.7.3) (2019-03-07)


### Performance Improvements

* tech release ([dddecc3](https://github.com/qiwi/common-formatters/commit/dddecc3))

## [0.7.2](https://github.com/qiwi/common-formatters/compare/v0.7.1...v0.7.2) (2019-03-01)


### Performance Improvements

* tech release ([b26b434](https://github.com/qiwi/common-formatters/commit/b26b434))

## [0.7.1](https://github.com/qiwi/common-formatters/compare/v0.7.0...v0.7.1) (2019-02-24)


### Bug Fixes

* **formatPercent:** repair default multiplier ([67f99ce](https://github.com/qiwi/common-formatters/commit/67f99ce))

# [0.7.0](https://github.com/qiwi/common-formatters/compare/v0.6.2...v0.7.0) (2019-02-24)


### Features

* expose validators and formatter defaults ([8b68283](https://github.com/qiwi/common-formatters/commit/8b68283))

## [0.6.2](https://github.com/qiwi/common-formatters/compare/v0.6.1...v0.6.2) (2019-02-21)


### Bug Fixes

* **package:** include typings to package ([63d08cb](https://github.com/qiwi/common-formatters/commit/63d08cb))

## [0.6.1](https://github.com/qiwi/common-formatters/compare/v0.6.0...v0.6.1) (2019-02-21)


### Bug Fixes

* throw away bundles from repo ([96459b8](https://github.com/qiwi/common-formatters/commit/96459b8))

# [0.6.0](https://github.com/qiwi/common-formatters/compare/v0.5.0...v0.6.0) (2019-02-21)


### Bug Fixes

* **lint:** repair lint script ([7e804bd](https://github.com/qiwi/common-formatters/commit/7e804bd))


### Features

* add libdefs generation ([83bc801](https://github.com/qiwi/common-formatters/commit/83bc801)), closes [#18](https://github.com/qiwi/common-formatters/issues/18)

# [0.5.0](https://github.com/qiwi/common-formatters/compare/v0.4.0...v0.5.0) (2018-12-03)


### Features

* add typographic minus symbol ([da6ed54](https://github.com/qiwi/common-formatters/commit/da6ed54))

# [0.4.0](https://github.com/qiwi/common-formatters/compare/v0.3.11...v0.4.0) (2018-09-18)


### Features

* **formatNumber:** add optional sign infication ([f382f3b](https://github.com/qiwi/common-formatters/commit/f382f3b)), closes [#17](https://github.com/qiwi/common-formatters/issues/17)
* **formatNumber:** support `fractionLength` option ([6504109](https://github.com/qiwi/common-formatters/commit/6504109))

lib/ts/phone.ts
import {
  IAny,
  IFormatted,
  IFormatter,
  IValidator
} from './interface'

import {clearNumericValue} from './util'

export const RUSSIAN_MOBILE_PHONE: string = '+* *** ***-**-**'

/**
 * @interface IFormatPhoneOpts
 * @property {string} blocksDelimiter
 * @property {string} countryCode
 * @property {string} areaCode
 * @property {boolean} areaBrackets
 * @property {string} countryCodePrefix
 * @property {number} countryCodeLength
 * @property {number} areaCodeLength
 * @property {number} phoneNumberLength
 * @property {string} phoneNumberDelimiter
 * @property {string} mask
 */
export type IFormatPhoneOpts = {
  strict?: boolean;
  blocksDelimiter?: string;
  countryCode?: string;
  areaCode?: string;
  areaBrackets?: boolean;
  countryCodePrefix?: string;
  countryCodeLength?: number;
  areaCodeLength?: number;
  phoneNumberLength?: number;
  phoneNumberDelimiter?: string;
  mask?: string | null
}

const DEFAULT_OPTS: IFormatPhoneOpts = {
  strict: true,
  countryCodePrefix: '+',
  blocksDelimiter: ' ',
  countryCodeLength: 0,
  areaCodeLength: 0,
  areaBrackets: false,
  phoneNumberDelimiter: '-',
  mask: null
}

export const validate: IValidator = (value: IAny) => !!value.length

/**
 * Phone formatter.
 * Relied on E.164 https://en.wikipedia.org/wiki/E.164
 *
 * @name formatPhone
 * @param {string} value
 * @param {IFormatPhoneOpts} [opts]
 * @return {string}
 * @example

 // Basic cases
 formatPhone('1234567')     // 123-45-67
 formatPhone('12345678')    // 1234-5678
 formatPhone('12345')       // 1-23-45
 formatPhone('1234567890')  // 1234567890

 // Format by mask
 formatPhone('79876543210', {mask: '+* *** ***-**-**'}) // +7 987 654-32-10

 // Format by opts
 formatPhone('223344', {countryCode: '7', areaCode: '8443', areaBrackets: true, phoneNumberDelimiter: '_'}) // +7 (8443) 22_33_44

 */
export const format: IFormatter = (value: IAny, opts?: IFormatPhoneOpts): IFormatted => {
  const cleared = clearNumericValue(value)
  const {
    strict,
    areaCode,
    countryCode,
    countryCodeLength,
    areaCodeLength,
    phoneNumberLength,
    phoneNumberDelimiter,
    blocksDelimiter,
    countryCodePrefix,
    areaBrackets,
    mask
  } = Object.assign({}, DEFAULT_OPTS, opts)

  if (strict && !validate(cleared)) {
    throw new Error('formatPhone: invalid input')
  }

  if (mask) {
    return formatByMask(cleared, mask)
  }

  const [rCountry, rArea, rPhone] = parseBlocks(cleared, countryCodeLength, areaCodeLength, phoneNumberLength)

  const country = formatCountryCode(rCountry || countryCode, countryCodePrefix)
  const area = formatAreaCode(rArea || areaCode, areaBrackets)
  const phone = formatPhoneNumber(rPhone, phoneNumberDelimiter)

  return [country, area, phone]
    .filter(v => v !== null)
    .join(blocksDelimiter)
}

export const parseBlocks = (value: string, ...blocks: Array<number|undefined>): Array<string> => {
  const lengths = resolveBlockLengths(value.length, ...blocks)
  let pos = 0

  return lengths.map((v: number) => {
    const substr = value.substr(pos, v)
    pos = pos + v
    return substr
  })
}

export const resolveBlockLengths = (entireLength: number, ...blocks: Array<number|undefined>): Array<number> => {
  const known = blocks.filter(v => v !== undefined) as Array<number>
  const sum: number = known.reduce((m, v) => m + v, 0)
  const diff = entireLength - sum

  switch (blocks.length - known.length) {
    case 0:
      if (diff !== 0) {
        throw new Error('formatPhone: invalid opts')
      }

      return known

    case 1:
      if (diff < 0) {
        throw new Error('formatPhone: invalid opts')
      }
      return blocks.map(v => v === null || v === undefined ? diff : v)

    default:
      throw new Error('formatPhone: invalid opts')
  }
}

export const formatAreaCode = (value?: string, brackets?: boolean): string | null=> {
  if (!value || !value.length) {
    return null
  }

  return brackets
    ? '(' + value + ')'
    : value
}

export const formatCountryCode = (value?: string, prefix?: string): string | null => {
  if (!value || !value.length) {
    return null
  }

  return prefix + value
}

export function formatPhoneNumber (value: string, delimiter: string = ''): string | null {
  switch (value.length) {
    case 0:
      return null

    case 5:
    case 6:
    case 7:
      return value.replace(/(.{1,3})(.{2})(.{2})/, '$1' + delimiter + '$2' + delimiter + '$3')

    case 8:
      return value.slice(0, 4) + delimiter + value.slice(4, 8)

    default:
      return value
  }
}

export function formatByMask (value: string, mask: string): string {
  const targetLen = mask.replace(/[^*]/g, '').length
  const values = value.split('')

  if (value.length !== targetLen) {
    throw new Error('formatPhone: input does not matches target mask')
  }

  // TODO optimize performance
  return mask
    .split('')
    .map(v => v === '*' ? values.shift() : v)
    .join('')
}

export default format
